src/v/storage/segment.cc

```diff
@@ -649,6 +649,7 @@ void segment::advance_stable_offset(size_t filepos) {
 
     _reader->set_file_size(it->first);
     _tracker.stable_offset = it->second;
+    _tracker.dirty_offset = std::max(_tracker.dirty_offset, it->second);
```

Comment:
You mentioned it offline, and I had the same thought while reviewing: it'd be great if the relationships between these were encapsulated in the methods that updated them, to avoid these kinds of bugs (doesn't have to be a part of this change though)